resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -557,4 +730,118 @@ private void addTrace(Throwable cause) {
+
+        // We not expect the linked-list to be very long so a double-linked-list is overkill.
+        private AuthoritativeNameServer head;
+        private int count;
+
+        AuthoritativeNameServerList(String questionName) {
+            this.questionName = questionName.toLowerCase(Locale.US);
+        }
+
+        void add(DnsRecord r) {
+            if (r.type() != DnsRecordType.NS || !(r instanceof DnsRawRecord)) {
+                return;
+            }
+
+            // Only include servers that serve the correct domain.
+            if (questionName.length() <  r.name().length()) {
+                return;
+            }
+
+            String recordName = r.name().toLowerCase(Locale.US);
+
+            int dots = 0;
+            for (int a = recordName.length() - 1, b = questionName.length() - 1; a >= 0; a--, b--) {
+                char c = recordName.charAt(a);
+                if (questionName.charAt(b) != c) {
+                    return;
+                }
+                if (c == '.') {
+                    dots++;
+                }
+            }
+
+            if (head != null && head.dots > dots) {
+                // We already have a closer match so ignore this one, no need to parse the domainName etc.
+                return;
+            }
+
+            final ByteBuf recordContent = ((ByteBufHolder) r).content();
+            final String domainName = decodeDomainName(recordContent);
+            if (domainName == null) {
+                // Could not be parsed, ignore.
+                return;
+            }
+
+            // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
+            // we drop servers that have a smaller dots count.
+            if (head == null || head.dots < dots) {
+                count = 1;
+                head = new AuthoritativeNameServer(dots, recordName, domainName);
+            } else if (head.dots == dots) {
+                AuthoritativeNameServer serverName = head;
+                while (serverName.next != null) {
+                    serverName = serverName.next;
+                }
+                serverName.next = new AuthoritativeNameServer(dots, recordName, domainName);
+                count++;
+            }
+        }
+
+        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
+        // one node less.
+        AuthoritativeNameServer remove(String nsName) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
+                    serverName.removed = true;
+                    return serverName;
+                }
+                serverName = serverName.next;
+            }
+            return null;
+        }
+
+        int size() {
+            return count;
+        }
+
+        static final class AuthoritativeNameServer {
```

Comment:
How about making this an inner class rather than an inner-inner class? I see no reason to nest this deep.